[docs] Add CI/CD tutorial (#44866)
expo/expo: docs/pages/tutorial/cicd/production.mdx

---
title: 'EAS Workflows for production deployments'
sidebar_title: Production deployments
description: Learn how to automate production deployments with EAS Workflows using fingerprint and release branches.

> ---

Reviewer: I'm not sure why but this page 404s in the preview https://pr-44866.expo-docs.pages.dev/tutorial/cicd/production/

Author: You need to clear your localstorage because of the project tracker component in the browser.

Let's add two jobs `update_android` and `update_ios` that run when there are no changes to native code and use `update` job type to trigger an OTA update:

```yaml .eas/workflows/production.yml
name: Deploy to production

on:
  push:
    branches: ['release/*']

jobs:
  fingerprint:
    # ...
  get_android_build:
    # ...
  get_ios_build:
    # ...
  build_android:
    # ...
  build_ios:
    # ...
  update_android:
    name: Publish Android update
    needs: [get_android_build]
    if: ${{ needs.get_android_build.outputs.build_id }}
    type: update
    params:
      branch: production
      platform: android
  update_ios:
    name: Publish iOS update
    needs: [get_ios_build]
    if: ${{ needs.get_ios_build.outputs.build_id }}
    type: update
    params:
      branch: production
      platform: ios
```

Our workflow only builds production-ready app binaries or publishes OTA updates.Automating store submission is a common next step discussed below in the [Automate app store submissions](#automate-app-store-submissions) section.

</Step>

<Step label="4">

### Create a release branch and push changes

> > **Note:** Ensure that the workflow file is already part of your GitHub repository and available on the `main` branch at this point.

Reviewer: Do we need to wrap this in an **info**?

---
title: 'Automate production deployments with EAS Workflows'
sidebar_title: Production deployments
description: Learn how to automate production builds and over-the-air updates with EAS Workflows from a release branch for Android and iOS.
---

import { ContentSpotlight } from '~/ui/components/ContentSpotlight';
import { BuildOrUpdate } from '~/ui/components/FlowDiagram/Diagrams/BuildOrUpdate';
import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { ProgressTracker } from '~/ui/components/ProgressTracker';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

> Rebuilding the app for every production change is wasteful when most changes are TypeScript/JavaScript only.

Reviewer 2: Suggest adding the chapter's resolution to the opener - the fingerprint decision is the marquee payoff here:

```suggestion
Rebuilding the app for every production change is wasteful when most changes are TypeScript/JavaScript only. By the end of this chapter, the production workflow builds only when native code changes and ships everything else as an over-the-air update.
```

## Learning outcomes

- Trigger production deployments from a `release/*` branch instead of every push to `main`

> - Branch between a native build and an [OTA update](/eas-update/introduction/) based on the project fingerprint

Reviewer 2: Small link retarget: for the concept term "OTA update", the concept page is a better first stop than the service intro (the page also defines the term):

```suggestion
- Branch between a native build and an [OTA update](/deploy/send-over-the-air-updates/) based on the project fingerprint
```

- Automate app store submissions as a follow-up step in the production workflow

<Prerequisites>

<Requirement title="EAS Build configured for production">

> [EAS CLI](/eas/cli/) prompts to generate credentials the first time we trigger a build for a profile. Trigger the production build manually for both Android and iOS so the credentials exist before the workflow runs using the command below:

Author: **[suggestion]** ~30-word sentence with a dangling `using the command below` (`expo-docs-style-guide.md#voice-and-tone`). Split into two sentences.

```suggestion
[EAS CLI](/eas/cli/) prompts to generate credentials the first time we trigger a build for a profile. Trigger the production build manually for both Android and iOS so the credentials exist before the workflow runs:
```

<Terminal cmd={['$ eas build --profile production --platform all']} />

</Requirement>

</Prerequisites>

## Using a release branch

Previously, with development builds, we have seen that using the `on.push.branches` trigger runs our workflow on every push to the specified `main` branch. For production deployments, triggering a workflow on every push to the `main` branch means every change merged is a new release. Teams usually want a deliberate release process, so they use a release branch or tag instead.

> In a release branch strategy, a branch is usually specified with a pattern such as `release/*`, where `*` is the name of the feature or version. This is also what separates Continuous Integration (CI) from Continuous Delivery (CD). CI runs on every push to the `main` branch, while CD runs on a release branch when the team is ready to deploy to production.

Author: **[suggestion]** Passive voice (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
In a release branch strategy, we usually specify the branch with a pattern such as `release/*`, where `*` is the feature or version name. This is what separates Continuous Integration (CI) from Continuous Delivery (CD). CI runs on every push to the `main` branch, while CD runs on a release branch when the team is ready to deploy to production.
```

| --------------------------------- | --------------------------------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| **.eas/workflows/build.yml**      | `on.push.branches: ['main']`      | `development` | CI workflow that runs on every push to the `main` branch. It can run unit tests and generate development builds.                                  |
| **.eas/workflows/preview.yml**    | `on.push.branches: ['main']`      | `preview`     | Preview workflow that runs on every push to the `main` branch. It can generate preview builds for internal testing and sharing with stakeholders. |
| **.eas/workflows/production.yml** | `on.push.branches: ['release/*']` | `production`  | CD workflow that runs on every push to a release branch.                                                                                          |

The above files coexist inside a project's GitHub repository without any conflict. Each uses a different build profile and a trigger.

<BuildOrUpdate />

## The `build` job type for production builds

> Let's start by creating a production workflow that uses the [pre-packaged jobs](/tutorial/cicd/first-workflow/#types-of-jobs-in-eas-workflows) `build` job type from EAS Workflows that runs for Android and iOS platforms.

Author: **[suggestion]** Three chained `that` clauses make the sentence harder than it needs to be (`expo-docs-style-guide.md#voice-and-tone`, one thought per sentence). Consider splitting.

```suggestion
Let's start by creating a production workflow. It uses the [pre-packaged](/tutorial/cicd/first-workflow/#types-of-jobs-in-eas-workflows) `build` job type from EAS Workflows and runs for both Android and iOS.
```

<Step label="1">

### Add a production.yml

Inside **.eas/workflows/**, create a new file called **production.yml** that uses the `production` build profile from **eas.json**:

```yaml .eas/workflows/production.yml
name: Deploy to production

on:
  push:
    branches: ['release/*']

jobs:
  build_android:
    name: Build Android
    type: build
    params:
      platform: android
      profile: production
  build_ios:
    name: Build iOS
    type: build
    params:
      platform: ios
      profile: production
```

> This will create a **.aab** artifact for Android and an **.ipa** artifact for iOS on a push to any branch that matches the pattern `release/*`.

Author: **[nit]** Present tense (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
This creates a **.aab** artifact for Android and an **.ipa** artifact for iOS on a push to any branch that matches the pattern `release/*`.
```

```yaml .eas/workflows/production.yml
name: Deploy to production

on:
  push:
    branches: ['release/*']

jobs:
  # @info #
  fingerprint:
    name: Fingerprint
    type: fingerprint
    environment: production
  # @end #
  # @info #
  get_android_build:
    name: Check for existing Android build
    needs: [fingerprint]
    type: get-build
    params:
      fingerprint_hash: ${{ needs.fingerprint.outputs.android_fingerprint_hash }}
      profile: production
  get_ios_build:
    name: Check for existing iOS build
    needs: [fingerprint]
    type: get-build
    params:
      fingerprint_hash: ${{ needs.fingerprint.outputs.ios_fingerprint_hash }}
      profile: production
  # @end #
  build_android:
    name: Build Android
    needs: [get_android_build]
    # @info #
    if: ${{ !needs.get_android_build.outputs.build_id }}
    # @end #
    type: build
    params:
      platform: android
      profile: production
  build_ios:
    name: Build iOS
    needs: [get_ios_build]
    # @info #
    if: ${{ !needs.get_ios_build.outputs.build_id }}
    # @end #
    type: build
    params:
      platform: ios
      profile: production
```

> In the above workflow, the `fingerprint` job runs first and generates a hash based on the changes in the codebase. The `get-build` jobs then check if there are existing builds for Android and iOS with the same fingerprint hash. If there are no existing builds, the `build` jobs will run to create new builds for production.

Author: **[nit]** Present tense (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
In the above workflow, the `fingerprint` job runs first and generates a hash based on the changes in the codebase. The `get-build` jobs then check if there are existing builds for Android and iOS with the same fingerprint hash. If there are no existing builds, the `build` jobs run to create new builds for production.
```

> The [`environment`](/eas/workflows/syntax/#jobsjob_idenvironment) field on the `fingerprint` job tells EAS which environment variables to load when computing the fingerprint. Our production environment may have different values for variables such as `API_URL` or feature flags, and those values can affect the native layer when EAS bakes them into the build. Setting `environment: production` ensures the fingerprint reflects what ships.

Author: **[suggestion]** Two thoughts joined by `and` (`expo-docs-style-guide.md#voice-and-tone`). The voice rule says two clauses joined by `and` is usually two sentences. Splitting makes the causal chain explicit.

```suggestion
The [`environment`](/eas/workflows/syntax/#jobsjob_idenvironment) field on the `fingerprint` job tells EAS which environment variables to load when computing the fingerprint. Our production environment may have different values for variables such as `API_URL` or feature flags. Those values can affect the native layer when EAS bakes them into the build. Setting `environment: production` ensures the fingerprint reflects what ships.
```

> The [`environment`](/eas/workflows/syntax/#jobsjob_idenvironment) field on the `fingerprint` job tells EAS which environment variables to load when computing the fingerprint. Our production environment may have different values for things like `API_URL` or feature flags, and those values can affect the native layer when they are baked into the build. Using `environment: production` makes sure the fingerprint reflects what ships.

Author: **[suggestion]** Passive `they are baked into` and `Using ... makes sure` can both be tightened (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
The [`environment`](/eas/workflows/syntax/#jobsjob_idenvironment) field on the `fingerprint` job tells EAS which environment variables to load when computing the fingerprint. Our production environment may have different values for variables such as `API_URL` or feature flags, and those values can affect the native layer when EAS bakes them into the build. Setting `environment: production` ensures the fingerprint reflects what ships.
```

> The [`environment`](/eas/workflows/syntax/#jobsjob_idenvironment) field on the `fingerprint` job tells EAS which environment variables to load when computing the fingerprint. Your production environment may have different values for things like `API_URL` or feature flags, and those values can affect the native layer when they're baked into the build. Using `environment: production` makes sure the fingerprint reflects what will ship.

Author: **[suggestion]** Tutorial pages use `we`/`our` throughout this PR; line 150 switches to `Your`/`they're` mid-paragraph (`expo-docs-style-guide.md#voice-and-tone`, tutorial exception). The same line also uses `what will ship` instead of present-tense `what ships`.

```suggestion
The [`environment`](/eas/workflows/syntax/#jobsjob_idenvironment) field on the `fingerprint` job tells EAS which environment variables to load when computing the fingerprint. Our production environment may have different values for things like `API_URL` or feature flags, and those values can affect the native layer when they are baked into the build. Using `environment: production` makes sure the fingerprint reflects what ships.
```

Let's add two jobs `update_android` and `update_ios` that run when there are no changes to native code and use `update` job type to trigger an OTA update:

```yaml .eas/workflows/production.yml
name: Deploy to production

on:
  push:
    branches: ['release/*']

jobs:
  fingerprint:
    # ...
  get_android_build:
    # ...
  get_ios_build:
    # ...
  build_android:
    # ...
  build_ios:
    # ...
  update_android:
    name: Publish Android update
    needs: [get_android_build]
    if: ${{ needs.get_android_build.outputs.build_id }}
    type: update
    params:
      branch: production
      platform: android
  update_ios:
    name: Publish iOS update
    needs: [get_ios_build]
    if: ${{ needs.get_ios_build.outputs.build_id }}
    type: update
    params:
      branch: production
      platform: ios
```

Our workflow only builds production-ready app binaries or publishes OTA updates. Automating store submission is a common next step discussed below in the [Automate app store submissions](#automate-app-store-submissions) section.

</Step>

<Step label="4">

### Create a release branch and push changes

> > **info** **Note:** Ensure that the workflow file is already part of your GitHub repository and available on the `main` branch at this point.

Author: **[nit]** Drop the redundant `**Note:**` after `**info**` (`expo-docs-components.md#callouts`). The line also breaks tutorial voice with `your GitHub repository`; rest of the chapter uses `our` (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
> **info** **Note:** Ensure that the workflow file is already part of our GitHub repository and available on the `main` branch at this point.
```

  src="/static/images/tutorial/cicd/eas-workflows-production-build-jobs.png"
  className="max-w-160"
/>

Both Android and iOS builds are running in parallel. Once each job completes, we receive a new artifact for Android and iOS.

Now, let's make a TypeScript/JavaScript only change in our example project on the same release branch and push again using the commands below:

<Terminal
  cmd={['$ git add .', '$ git commit -m "Update welcome text"', '$ git push origin release/1.0.0']}
  cmdCopy="git add . && git commit -m 'Update welcome text' && git push origin release/1.0.0"
/>

After the workflow runs, notice in the EAS dashboard that the fingerprint matches the build from the last push. The `build` jobs are skipped entirely, and the `update` jobs publish an OTA update.

<ContentSpotlight
  alt="EAS Workflows production run reusing existing native builds via fingerprint match and publishing Android and iOS over-the-air updates."
  src="/static/images/tutorial/cicd/eas-workflows-production-update-jobs.png"
  className="max-w-160"
/>

</Step>

## Automate app store submissions

> The workflow we have built in this chapter only builds production-ready app binaries or publishes OTA updates to an existing production app. Automating [store submission](/build/automate-submissions/) is a common next step in a production release workflow.

Reviewer 2: nit: the link covers only "store submission", leaving "Automating" dangling outside it - linking the whole phrase reads cleaner:

```suggestion
The workflow we have built in this chapter only builds production-ready app binaries or publishes OTA updates to an existing production app. [Automating store submission](/build/automate-submissions/) is a common next step in a production release workflow.
```

EAS Workflows provides a `submit` pre-packaged job for automating app store submissions. It requires us to manage our app store credentials with EAS CLI. It also requires us to fulfill app store requirements, such as manually uploading the first Android release (**.aab**) to the Google Play Store. iOS has its own Apple Developer Program enrollment and signing credentials to set up.

Once the app store requirements are in place, we can add two new jobs `submit_android` and `submit_ios` to our workflow that run after the `build` jobs:

```yaml .eas/workflows/production.yml
# rest of the workflow file

submit_android:
  name: Submit Android
  needs: [build_android]
  type: submit
  params:
    build_id: ${{ needs.build_android.outputs.build_id }}
submit_ios:
  name: Submit iOS
  needs: [build_ios]
  type: submit
  params:
    build_id: ${{ needs.build_ios.outputs.build_id }}
```

> The above `submit_android` and `submit_ios` jobs require a `build_id` parameter to know which build to submit to the app stores. They are going to run only when the `build_android` and `build_ios` jobs run, which means a new production build is created. If there are only changes to TypeScript/JavaScript files, the `submit` jobs are skipped as well since no new build is created.

Author: **[nit]** Present tense; `are going to run` is the future construction (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
The above `submit_android` and `submit_ios` jobs require a `build_id` parameter to know which build to submit to the app stores. They run only when the `build_android` and `build_ios` jobs run, which means a new production build is created. If there are only changes to TypeScript/JavaScript files, the `submit` jobs are skipped as well since no new build is created.
```

## Summary

<ProgressTracker
  currentChapterIndex={4}
  name="CICD_TUTORIAL"

>   summary="We created a production workflow for Android and iOS triggered by pushes to a release branch, used fingerprinting to choose between a native build and an OTA update, and saw how EAS Submit extends the workflow to automate app store submissions."

Author: **[suggestion]** ProgressTracker `summary` is 41 words (`expo-docs-style-guide.md#voice-and-tone`). Parallel three-part structure reads, but this is the longest summary in the tutorial. Tightening the third clause matches the other chapter summaries.

```suggestion
  summary="We created a production workflow for Android and iOS triggered by pushes to a release branch, used fingerprinting to choose between a native build and an OTA update, and saw how EAS Submit automates app store submissions."
```
